### DISCUSSION TITLE
Upgrading to new version

### DISCUSSION THREAD

**Discussion by Dhuel** [litedb-org/LiteDB#2046] [Q&A]
When upgrading from an old version of the package, new errors are being thrown around the delete function.

The old layout accepted 

public virtual T DeleteItemAsync(T item)
    {
        var c = _collection.Delete(i => i.Equals(item));
        return item;
    }

now,, there is an error which states "cannot convert lambda expression to type 'BsonValue' because it is not a delagate type

**Comment by Dhuel** [comment]
I believe i figured it out. 

var doc = BsonMapper.Global.ToDocument(customer);

**Comment by lbnascimento** [comment]
On older versions of LiteDB, lambda expressions were applied after deserialization. This was very flexible, but it had poor performance – for instance, if you ran something like `collection.Delete(x => x.Equals(myItem))`, every single document in the collection would have to be retrieved, deserialized and applied to the expression just to figure out whether it should be deleted or not.

This was changed in v5. Lambda expressions are now converted to `BsonExpression`s, which are expressions that can be directly applied to a `BsonDocument`. This is less flexible because you can't use arbitrary expressions, but it's much faster. 

The recommended way to delete documents is using their ids. You can also use `DeleteMany`(...) with *some* lambda expression, like `collection.DeleteMany(x => x.Name.StartsWith("John"))`, but this can't be done in a generic method.

If you absolutely must use a generic method and the comparison logic implemented in `Equals(...)`, you have to implement it manually.
